### PR TITLE
build: Update min version to k8s 1.17

### LINKED ISSUE #10115: Update min K8s version supported to 1.17
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
K8s 1.16 is currently the min version supported by Rook, but it has the following limitations:
- CSI has a number of limitations such as snapshots are not stable
- Priority classes are not supported

The proposal is to remove support for K8s 1.16 in the next Rook v1.9.x release. 
- Development will be simplified since the CI does not need to run on that version. 
- The min version tested by the CI will now be v1.17.x
- Rook is not prevented from running on K8s 1.16 and may certainly continue running there, it is just not tested by the CI

### PR DESCRIPTION
**Pull request by travisn**
**Description of your changes:**
Support is removed for k8s for various limitations such as priority classes not working and csi driver feature incompleteness and missing snapshots. Documentation is updated with the new min version of k8s 1.17 and also the tests are updated to run on the min version of 1.17.

Let's hold this for a few days to wait for feedback on #10115. 

**Which issue is resolved by this Pull Request:**
Resolves #10115 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by travisn** [comment]
No negative feedback on #10115 so let's go ahead with this

File: pkg/operator/ceph/cluster/osd/osd_test.go [hunk 1]
```diff
@@ -565,10 +565,10 @@ func TestDetectCrushLocation(t *testing.T) {
...
 	// update the location with valid topology labels
 	nodeLabels = map[string]string{
-		"failure-domain.beta.kubernetes.io/region": "region1",
-		"failure-domain.beta.kubernetes.io/zone":   "zone1",
-		"topology.rook.io/rack":                    "rack1",
-		"topology.rook.io/row":                     "row1",
+		"topology.kubernetes.io/region": "region1",
+		"topology.kubernetes.io/zone":   "zone1",
+		"topology.rook.io/rack":         "rack1",
+		"topology.rook.io/row":          "row1",
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/osd/osd_test.go:571]: We should leave these checks. Even though the label is deprecated, support is not fully removed. In K8s, "deprecated" is not synonymous with "unsupported." Kubernetes docs still include them in their well-known labels list: https://kubernetes.io/docs/reference/labels-annotations-taints/

File: pkg/operator/ceph/cluster/osd/topology.go [hunk 2]
```diff
@@ -63,31 +63,17 @@ func extractTopologyFromLabels(labels map[string]string) (map[string]string, str
...
 	const zoneLabel = "zone"
-	zone, ok := labels[corev1.LabelZoneFailureDomain]
-	if ok {
-		topology[zoneLabel] = zone
-		topologyAffinity = formatTopologyAffinity(corev1.LabelZoneFailureDomain, zone)
-	}
-
-	// check for the zone k8s topology label that is GA in 1.17.
-	zone, ok = labels[corev1.LabelZoneFailureDomainStable]
+	zone, ok := labels[corev1.LabelZoneFailureDomainStable]
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/osd/topology.go:76]: Ditto. Even though the label is deprecated, support is not fully removed. Kubernetes docs still include them in their well-known labels list: https://kubernetes.io/docs/reference/labels-annotations-taints/

File: pkg/operator/ceph/cluster/osd/topology_test.go [hunk 3]
```diff
@@ -105,14 +105,14 @@ func TestTopologyLabels(t *testing.T) {
...
-		corev1.LabelZoneFailureDomain: "z1",
+		corev1.LabelZoneRegionStable:        "r1",
+		corev1.LabelZoneFailureDomainStable: "z1",
 	}
 	topology, affinity = extractTopologyFromLabels(nodeLabels)
 	assert.Equal(t, 2, len(topology))
 	assert.Equal(t, "r1", topology["region"])
 	assert.Equal(t, "z1", topology["zone"])
-	assert.Equal(t, "failure-domain.beta.kubernetes.io/zone=z1", affinity)
+	assert.Equal(t, "topology.kubernetes.io/zone=z1", affinity)
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/osd/topology_test.go:115]: Ditto again, the deprecated labels are not "unsupported" labels.

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
There is really just one issue I see which is the misunderstanding that the "deprecated" labels `failure-domain.beta.kubernetes.io/*` are "unsupported". Kubernetes docs still reference this label, and Kubernetes follows the trend that deprecation comes before removal of support. I think we will have to perpetually support these labels for quite some time since these labels are also tied to other utilities, and users may set these labels themselves for quite some time.

**Review by BlaineEXE — APPROVED** [review]
Looks good to me.

I wonder if it would be good to have a check at the top level of the operator to fail if the kube version isn't high enough (1.17 for now). Pro: It could prevent people from upgrading Rook on a K8s cluster that is too old, saving them from using an unsupported version combination. Con: It would prevent users from using the combination unnecessarily if they don't use any of the 1.17+ features and feel confident about what they're doing.

**Comment by travisn** [comment]
> I wonder if it would be good to have a check at the top level of the operator to fail if the kube version isn't high enough (1.17 for now). Pro: It could prevent people from upgrading Rook on a K8s cluster that is too old, saving them from using an unsupported version combination. Con: It would prevent users from using the combination unnecessarily if they don't use any of the 1.17+ features and feel confident about what they're doing.

I could see logging a warning, but the philosophy so far has not been to block running on older K8s versions, also mentioned briefly in #10115. Documenting the min version has so far been sufficient.

**Pull request merged**